linker/Linker.Steps/MarkStep.cs

```diff
@@ -480,6 +480,11 @@ protected virtual bool ShouldMarkTopLevelCustomAttribute (AttributeProviderPair
 			// then surely nothing is using this attribute and there is no need to mark it
 			if (!Annotations.IsMarked (resolvedConstructor.Module) && !Annotations.IsMarked (ca.AttributeType))
 				return false;
+
+			// If module the custom attribute appears in has not been marked yet, then don't mark the attribute
+			// i.e Don't mark custom attributes on an assembly until the assembly has been marked
+			if (!Annotations.IsMarked (ca.AttributeType.Module))
```

Comment:
The check before this is

`if (!Annotations.IsMarked (resolvedConstructor.Module) && !Annotations.IsMarked (ca.AttributeType))`

I.e. If the resolved module is not marked and the attribute TypeReference is not marked.

Imagine `[assembly: Debuggable]` in an assembly that we are going to strip away because nothing is used.

`Annotations.IsMarked (resolvedConstructor.Module) ` is going to be true because `resolvedConstructor.Module` is mscorlib.

With this new if `if (!Annotations.IsMarked (ca.AttributeType.Module))` , `Annotations.IsMarked (ca.AttributeType.Module)` is going to be false because `ca.AttributeType.Module` is our assembly that we can strip away and it has not been marked

Maybe, the two if's can be combine.  I didn't try and think about it for a few reasons

* This is kind of delicate edge case stuff and I don't want to introduce a regression

* I can't (easily) write a test in the upstream linker to cover this new if I'm adding because the assembly is going to be deleted anyways.  We'd need this "no delete" mode I'm adding to our UnityLinker in order to test this.  As a result, I'd rather the logic be it's own dedicated if rather than mixed into some other if where it's purpose might be less clear.

Reply:
I don't like adding yet another combination without understanding existing logic.

We already have a case which does same thing here https://github.com/mono/linker/blob/44f3b1e740b7156b7109d04a9f58a0473c78a466/linker/Linker.Steps/MarkStep.cs#L465 but under KeepUsedAttributeTypesOnly. This kind of removes the condition.

Secondly, I am not sure why do you have `.Module` there. It basically blocks any attributes declared in mscorlib to be ever stripped when referenced.

Reply:
The goal of the new if I'm adding is to not mark `CustomAttribute` instances on the Assembly when the Assembly is going to be deleted regardless of whether `KeepUsedAttributeTypesOnly` is enabled or not.

The goal of the if you pointed out is to not mark `CustomAttribute` instances until the TypeDefinition of the AttributeType has been marked.

> Secondly, I am not sure why do you have .Module there. It basically blocks any attributes declared in mscorlib to be ever stripped when referenced.

I don't follow what your concern is.  I added a change set to this PR to support asserting attributes in other assemblies and I added new tests to assert what happens with attributes defined in mscorlib.  Nothing seems wrong..
https://github.com/mono/linker/pull/360/commits/7a08b2e51b466bae6e4ff3accd9d91497553056f